pennylane/labs/dla/lie_closure_dense.py

```diff
@@ -0,0 +1,156 @@
+
+
+def _hermitian_basis(matrices: Iterable[np.ndarray], tol: float = None, subbasis_length: int = 0):
+    """Find a linearly independent basis of a list of (skew-) Hermitian matrices
+
+    Args:
+        matrices (Iterable[numpy.ndarray]): A list of Hermitian matrices.
+        tol (float): Tolerance for linear dependence check. Defaults to ``1e-10``.
+
+    Returns:
+        np.ndarray: Stacked array of linearly independent basis matrices.
+
+    Raises:
+        ValueError: If not all input matrices are (skew-) Hermitian.
+    """
+    if tol is None:
+        tol = 1e-10
+
+    basis = list(matrices[:subbasis_length])
+    for A in matrices[subbasis_length:]:
+        if not np.allclose(A.conj().T, A):
+            A = 1j * A
+            if not np.allclose(A.conj().T, A):
+                raise ValueError(f"At least one basis matrix is not (skew-)Hermitian:\n{A}")
+
+        B = A.copy()
+        if len(basis) > 0:
+            B -= np.tensordot(
+                np.tensordot(np.array(basis).conj(), A, axes=[[1, 2], [0, 1]]),
+                basis,
+                axes=[[0], [0]],
+            )
+        if (norm := np.linalg.norm(B)) > tol:  # Tolerance for numerical stability
+            B /= norm
+            basis.append(B)
+    return np.array(basis)
+
+
+def lie_closure_dense(
+    generators: Iterable[Union[PauliWord, PauliSentence, Operator]],
+    n=None,
+    max_iterations: int = 10000,
+    verbose: bool = False,
+    tol: float = None,
+):
+    r"""Compute the dynamical Lie algebra :math:`\mathfrak{g}` from a set of generators using their dense matrix representation.
+
+    This function computes the Lie closure of a set of generators using their dense matrix representation.
+    This is sometimes more efficient than using the sparse Pauli representations of :class:`~PauliWord` and
+    :class:`~PauliSentence` that are employed in :func:`~lie_closure`, e.g., when there are few generators
+    that are sums of many Paulis.
+
+    .. seealso:: For details on the mathematical definitions, see :func:`~lie_closure` and our `Introduction to Dynamical Lie Algebras for quantum practitioners <https://pennylane.ai/qml/demos/tutorial_liealgebra/>`__.
+
+    Args:
+        generators (Iterable[Union[PauliWord, PauliSentence, Operator]]): generating set for which to compute the
+            Lie closure.
+        n (int): The number of qubits involved. If ``None`` is provided, it is automatically deduced from the generators.
+        max_iterations (int): maximum depth of nested commutators to consider. Default is ``10000``.
+        verbose (bool): whether to print out progress updates during Lie closure
+            calculation. Default is ``False``.
+        tol (float): Numerical tolerance for the linear independence check used in :class:`~.PauliVSpace`.
+
+    Returns:
+        numpy.ndarray: The ``(dim(g), 2**n, 2**n)`` array containing the linear independent basis of the DLA g as dense matrices.
+
```

Comment:
Could include a "Raises" section for the assertions we're doing :) Also, should they be properly raise ValueErrors or something?

Reply:
Probably yes, will fix

Reply:
Actually not sure here since those are just consistency checks, dont know what the rules are for those, but I recall seeing this in other places in the codebase 🤔

Reply:
As this is the `labs` module, I'm fine with both :)